File: src/org/antlr/runtime/tree/DOTTreeGenerator.java
/*
 [The "BSD licence"]
 Copyright (c) 2005-2006 Terence Parr
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package org.antlr.runtime.tree;

import org.antlr.stringtemplate.StringTemplate;

/** A utility class to generate DOT diagrams (graphviz) from
 *  arbitrary trees.  You can pass in your own templates and
 *  can pass in any kind of tree or use Tree interface method.
 *  I wanted this separator so that you don't have to include
 *  ST just to use the org.antlr.runtime.tree.* package.
 *  This is a set of non-static methods so you can subclass
 *  to override.  For example, here is an invocation:
 *
 *      CharStream input = new ANTLRInputStream(System.in);
 *      TLexer lex = new TLexer(input);
 *      CommonTokenStream tokens = new CommonTokenStream(lex);
 *      TParser parser = new TParser(tokens);
 *      TParser.e_return r = parser.e();
 *      Tree t = (Tree)r.tree;
 *      System.out.println(t.toStringTree());
 *      DOTTreeGenerator gen = new DOTTreeGenerator();
 *      StringTemplate st = gen.toDOT(t);
 *      System.out.println(st);
 */
public class DOTTreeGenerator {

    public static StringTemplate _treeST =
        new StringTemplate(
            "digraph {\n" +
            "  node [shape=plaintext, fixedsize=true, fontsize=11, fontname=\"Courier\",\n" +
            "        width=.4, height=.2];\n" +
            "  edge [arrowsize=.7]\n" +
            "  $nodes$\n" + // only used when edges empty; single node tree
            "  $edges$\n" +
            "}\n");

    public static StringTemplate _edgeST =
            new StringTemplate("\"$parent$\" -> \"$child$\"\n");

    public StringTemplate toDOT(Object tree,
                                TreeAdaptor adaptor,
                                StringTemplate _treeST,
                                StringTemplate _edgeST)
    {
        StringTemplate treeST = _treeST.getInstanceOf();
        toDOTTreeWork(tree, adaptor, treeST);
        if ( adaptor.getChildCount(tree)==0 ) {
            // single node, don't do edge.
            treeST.setAttribute("nodes", adaptor.getText(tree));
        }
        return treeST;
    }

    public StringTemplate toDOT(Object tree,
                                TreeAdaptor adaptor)
    {
        return toDOT(tree, adaptor, _treeST, _edgeST);
    }

    /** Generate DOT (graphviz) for a whole tree not just a node.
     *  For example, 3+4*5 should generate:
     *
     * digraph {
     *   node [shape=plaintext, fixedsize=true, fontsize=11, fontname="Courier",
     *         width=.4, height=.2];
     *   edge [arrowsize=.7]
     *   "+"->3
     *   "+"->"*"
     *   "*"->4
     *   "*"->5
     * }
     *
     * Return the ST not a string in case people want to alter.
     *
     * Takes a Tree interface object.
     */
    public StringTemplate toDOT(Tree tree) {
        return toDOT(tree, new CommonTreeAdaptor());
    }

    protected void toDOTTreeWork(Object tree,
                                 TreeAdaptor adaptor,
                                 StringTemplate treeST)
    {
        if ( tree==null ) {
            return;
        }
        int n = adaptor.getChildCount(tree);
        if ( n==0 ) {
            // must have already dumped as child from previous
            // invocation; do nothing
            return;
        }

        // for each child, do a parent -> child edge
        String parentText = adaptor.getText(tree);
        for (int i = 0; i < n; i++) {
            Object child = adaptor.getChild(tree, i);
            String childText = adaptor.getText(child);
            StringTemplate edgeST = _edgeST.getInstanceOf();
            edgeST.setAttribute("parent", parentText);
            edgeST.setAttribute("child", childText);
            treeST.setAttribute("edges", edgeST);
            toDOTTreeWork(child, adaptor, treeST);
        }
    }

}

File: runtime/CSharp/Antlr3.Utility/Antlr.Utility.Tree/DOTTreeGenerator.cs
namespace Antlr.Utility.Tree
{
    using Antlr.Runtime.Tree;
    using Antlr.StringTemplate;
    using System;
    using System.Collections;

    public class DOTTreeGenerator
    {
        public static Antlr.StringTemplate.StringTemplate _edgeST = new Antlr.StringTemplate.StringTemplate("$parent$ -> $child$ // \"$parentText$\" -> \"$childText$\"\n");
        public static Antlr.StringTemplate.StringTemplate _nodeST = new Antlr.StringTemplate.StringTemplate("$name$ [label=\"$text$\"];\n");
        public static Antlr.StringTemplate.StringTemplate _treeST = new Antlr.StringTemplate.StringTemplate("digraph {\n  ordering=out;\n  ranksep=.4;\n  node [shape=plaintext, fixedsize=true, fontsize=11, fontname=\"Courier\",\n        width=.25, height=.25];\n  edge [arrowsize=.5]\n  $nodes$\n  $edges$\n}\n");
        private int nodeNumber;
        private IDictionary nodeToNumberMap = new Hashtable();

        protected int GetNodeNumber(object t)
        {
            object obj2 = this.nodeToNumberMap[t];
            if (obj2 != null)
            {
                return (int) obj2;
            }
            this.nodeToNumberMap[t] = this.nodeNumber;
            this.nodeNumber++;
            return (this.nodeNumber - 1);
        }

        protected Antlr.StringTemplate.StringTemplate GetNodeST(ITreeAdaptor adaptor, object t)
        {
            string nodeText = adaptor.GetNodeText(t);
            Antlr.StringTemplate.StringTemplate instanceOf = _nodeST.GetInstanceOf();
            string val = "n" + this.GetNodeNumber(t);
            instanceOf.SetAttribute("name", val);
            instanceOf.SetAttribute("text", nodeText);
            return instanceOf;
        }

        public Antlr.StringTemplate.StringTemplate ToDOT(ITree tree)
        {
            return this.ToDOT(tree, new CommonTreeAdaptor());
        }

        public Antlr.StringTemplate.StringTemplate ToDOT(object tree, ITreeAdaptor adaptor)
        {
            return this.ToDOT(tree, adaptor, _treeST, _edgeST);
        }

        public Antlr.StringTemplate.StringTemplate ToDOT(object tree, ITreeAdaptor adaptor, Antlr.StringTemplate.StringTemplate _treeST, Antlr.StringTemplate.StringTemplate _edgeST)
        {
            Antlr.StringTemplate.StringTemplate instanceOf = _treeST.GetInstanceOf();
            this.nodeNumber = 0;
            this.ToDOTDefineNodes(tree, adaptor, instanceOf);
            this.nodeNumber = 0;
            this.ToDOTDefineEdges(tree, adaptor, instanceOf);
            return instanceOf;
        }

        protected void ToDOTDefineEdges(object tree, ITreeAdaptor adaptor, Antlr.StringTemplate.StringTemplate treeST)
        {
            if (tree != null)
            {
                int childCount = adaptor.GetChildCount(tree);
                if (childCount != 0)
                {
                    string val = "n" + this.GetNodeNumber(tree);
                    string nodeText = adaptor.GetNodeText(tree);
                    for (int i = 0; i < childCount; i++)
                    {
                        object child = adaptor.GetChild(tree, i);
                        string str3 = adaptor.GetNodeText(child);
                        string str4 = "n" + this.GetNodeNumber(child);
                        Antlr.StringTemplate.StringTemplate instanceOf = _edgeST.GetInstanceOf();
                        instanceOf.SetAttribute("parent", val);
                        instanceOf.SetAttribute("child", str4);
                        instanceOf.SetAttribute("parentText", nodeText);
                        instanceOf.SetAttribute("childText", str3);
                        treeST.SetAttribute("edges", instanceOf);
                        this.ToDOTDefineEdges(child, adaptor, treeST);
                    }
                }
            }
        }

        protected void ToDOTDefineNodes(object tree, ITreeAdaptor adaptor, Antlr.StringTemplate.StringTemplate treeST)
        {
            if (tree != null)
            {
                int childCount = adaptor.GetChildCount(tree);
                if (childCount != 0)
                {
                    Antlr.StringTemplate.StringTemplate nodeST = this.GetNodeST(adaptor, tree);
                    treeST.SetAttribute("nodes", nodeST);
                    for (int i = 0; i < childCount; i++)
                    {
                        object child = adaptor.GetChild(tree, i);
                        Antlr.StringTemplate.StringTemplate val = this.GetNodeST(adaptor, child);
                        treeST.SetAttribute("nodes", val);
                        this.ToDOTDefineNodes(child, adaptor, treeST);
                    }
                }
            }
        }
    }
}

